framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -366,6 +378,119 @@ public S initialStore(UnderlyingAST underlyingAST, List<LocalVariableNode> param
     return store;
   }
 
+  /**
+   * Determine whether a lambda expression is leaked.
+   *
+   * <p>This currently checks for the {@link NonLeaked} annotation, which is trusted.
+   *
+   * @param lambda the lambda expression
+   * @param aTypeFactory an annotated type factory
+   * @return true if the lambda expression is found to be annotated with {@link NonLeaked}
+   */
+  private boolean isLambdaLeaked(CFGLambda lambda, AnnotatedTypeFactory aTypeFactory) {
+    LambdaExpressionTree lambdaTree = lambda.getLambdaTree();
+    TreePath lambdaPath = aTypeFactory.getPath(lambdaTree);
+    Tree lambdaParent = lambdaPath.getParentPath().getLeaf();
+    if (lambdaParent.getKind() == Tree.Kind.METHOD_INVOCATION) {
+      MethodInvocationTree invok = (MethodInvocationTree) lambdaParent;
+      ExecutableElement methodElt = TreeUtils.elementFromUse(invok);
+      AliasingAnnotatedTypeFactory aliasingAtf =
+          analysis
+              .atypeFactory
+              .getChecker()
+              .getTypeFactoryOfSubcheckerOrNull(AliasingChecker.class);
+      if (aliasingAtf != null) {
+        // Check whether any parameter is NOT annotated with @NonLeaked
+        return methodElt.getParameters().stream()
+            .anyMatch(
+                param ->
+                    aliasingAtf.getAnnotatedType(param).getPrimaryAnnotation(NonLeaked.class)
+                        == null);
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns true if a given lambda expression is pure. A lambda expression is pure if:
+   *
+   * <ol>
+   *   <li>All methods invoked within the lambda expression are known to be pure.
```

Comment:
This isn't enough, because the lambda itself might perform assignments, making it impure.  As a subcase of that, it could include increment/decrement operations like i++ or --j.
I think the implementation requires that the body be of a specific form -- a chain of method calls -- that makes this adequate.